domain/src/main/java/org/oppia/domain/oppialogger/logstorage/LogReportStorageHelper.kt

```diff
@@ -0,0 +1,172 @@
+package org.oppia.domain.oppialogger.logstorage
+
+import androidx.lifecycle.LiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import org.oppia.app.model.EventLog
+import org.oppia.app.model.ExceptionLog
+import org.oppia.app.model.OppiaEventLogs
+import org.oppia.app.model.OppiaExceptionLogs
+import org.oppia.data.persistence.PersistentCacheStore
+import org.oppia.domain.oppialogger.EventLogStorageCacheSize
+import org.oppia.domain.oppialogger.ExceptionLogStorageCacheSize
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.data.DataProviders
+import org.oppia.util.logging.ConsoleLogger
+import org.oppia.util.threading.BackgroundDispatcher
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Helper class for storing log reports in absence of network connectivity. */
+@Singleton
+class LogReportStorageHelper @Inject constructor(
+  cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val consoleLogger: ConsoleLogger,
+  @BackgroundDispatcher private val backgroundCoroutineDispatcher: CoroutineDispatcher,
+  @EventLogStorageCacheSize private val eventLogStorageCacheSize: Int,
+  @ExceptionLogStorageCacheSize private val exceptionLogStorageCacheSize: Int
+) {
+  private enum class LogReportType {
+    EVENT_LOG,
+    EXCEPTION_LOG
+  }
+  private val eventLogStore =
+    cacheStoreFactory.create("event_logs", OppiaEventLogs.getDefaultInstance())
+  private val exceptionLogStore =
+    cacheStoreFactory.create("exception_logs", OppiaExceptionLogs.getDefaultInstance())
+  private val coroutineScope = CoroutineScope(backgroundCoroutineDispatcher)
+
+  /** Adds an event to the storage. */
+  fun addEventLog(eventLog: EventLog) {
+    coroutineScope.launch {
+      checkStoreCacheStatus(
+        eventLogStore.readDataAsync().await().eventLogList.size + 1,
+        LogReportType.EVENT_LOG,
+        eventLogStorageCacheSize
+      )
+    }
+    eventLogStore.storeDataAsync(updateInMemoryCache = true) {
+      it.toBuilder().addEventLog(eventLog).build()
+    }.invokeOnCompletion {
+      it?.let {
+        consoleLogger.e(
+          "DOMAIN",
+          "Failed to store event log",
+          it
+        )
+      }
+    }
+  }
+
+  /** Adds an exception to the storage. */
+  fun addExceptionLog(exceptionLog: ExceptionLog) {
+    coroutineScope.launch {
```

Comment:
Performing operations that need to be synchronized on the internal state of a cache should be done within that cache's class, if possible. Otherwise you're risking introducing complex race conditions (in this case, I think it's possible for logs to exceed the maximum number expected in an accumulating manner).

Reply:
Modified accordingly, PTAL :)